### ISSUE TITLE
Review of MLR metrics

### ISSUE THREAD

**Issue by tlienart** [alan-turing-institute/MLJBase.jl#95]
https://mlr.mlr-org.com/articles/tutorial/measures.html

### classification

MLR | MLJBase | Comment
---- | --- | ---
acc | accuracy | 
auc | auc | 
bac | **missing** | 
ber | **missing** |
brier | BrierScore |
brier scaled | **missing?** | maybe worth checking?
f1 | f1score |
fdr | **missing** | #88 
fn | **missing** | #88 
fnr | **missing** | #88 
fp | **missing** #88 
fpr | **missing** | #88 
gmean | **missing** | 
gpr | **missing** |
kappa | **missing** |
logloss | cross entropy ?  | **check**
lsr | **missing** |
mcc | **missing** |
mmce | **missing** |
multiclass au1p | **missing** | 👀 
multiclass au1u | **missing** |
multiclass aunp | **missing** |
mutliclass aunu | **missing** |
multiclass brier | **missing** |
npv | **missing** | #88 
ppv | **missing** | #88 
qsr | **missing** | 
ssr | **missing** |
tn | **missing** | #88 
tnr | **missing** | #88 
tp | **missing** | #88 
tpr | **missing** | #88 
wkappa | **missing** | #88 

### regression

some of these may be available in LossFunctions (?) + I did this one on the top of my head so may be worth double checking

MLR | MLJBase | Comment
---- | --- | ---
arsq | **missing** |
expvar | **missing** |
kendalltau | **missing** |
mae | mav | **check**
mape | **missing** | 
medae | **missing** |
medse | **missing** |
mse | mse |
msle | **missing** |
rae | **missing** |
rmse | rms | 
rmsle | rmsl |
rrse | **missing** |
rsq | **missing** |
sae | **missing** |
spearmanrho | **missing** |
sse | **missing** |

**Comment by tlienart** [comment]
Might be good to also add stuff like rsquared, SSE, TSS etc for regression metrics

**Comment by azev77** [comment]
@tlienart I submitted a [PR](https://github.com/alan-turing-institute/MLJBase.jl/pull/293/files) to add MAPE. 

Copying #294:
Rob's paper: https://robjhyndman.com/papers/mase.pdf (might be a newer version)
has a nice review of different measures of forecast accuracy.

Would be nice to incorporate at some point.
(They focus on univariate time series, but their results apply more broadly.)

2.1 Scale-dependent measures:   **e := ŷ - y**
Mean Square Error (MSE)
Root Mean Square Error (RMSE)
Mean Absolute Error (MAE)
Median Absolute Error (MdAE)

2.2 Measures based on percentage errors:    **p := 100*e/y**
Mean Absolute Percentage Error (MAPE)
Median Absolute Percentage Error (MdAPE)
Root Mean Square Percentage Error (RMSPE)
Root Median Square Percentage Error (RMdSPE)
Symmetric Mean Absolute Percentage Error (sMAPE)
Symmetric Median Absolute Percentage Error (sMdAPE)

2.3 Measures based on relative errors (Rel to benchmark) **r := e/e*** 
Mean Relative Absolute Error (MRAE)
Median Relative Absolute Error (MdRAE)
Geometric Mean Relative Absolute Error (GMRAE)

2.4 Relative measures. {is Rsquared in this category?}   
RelMAE=MAE/MAEb

3 Scaled errors
Mean Absolute Scaled Error (MASE)

Check out his TS book too: https://otexts.com/fpp2/accuracy.html

**Comment by azev77** [comment]
& this unmaintained repo has plenty of Julia coded measures: https://github.com/JuliaML/MLMetrics.jl
Along w/ Flux: https://github.com/FluxML/Flux.jl/blob/0287abbf663b1ae276360ea4a396adad1d2d1df7/src/layers/stateless.jl

**Comment by ablaom** [comment]
I've looked at the flux measures and I think their API is a little different because they want the measures to work on batches. And they only have a few. And I've had problems getting them to work as expected.

**Comment by azev77** [comment]
```julia
function err(ŷ, y)
    return ŷ - y
end

function perr(ŷ, y; tol = eps())
    e = err(ŷ, y)
    p = 100 * e[y .!= 0] ./ y[y .!= 0]
    return p
end

using Statistics
mse(ŷ, y)  = err(ŷ, y) |> (x)->x.^2 |> mean
rmse(ŷ, y) = err(ŷ, y) |> x -> x.^2 |> mean |> sqrt
mae(ŷ, y)  = err(ŷ, y) |> x -> abs.(x) |> mean
mdae(ŷ, y) = err(ŷ, y) |> x -> abs.(x) |> median
rmdse(ŷ, y) = err(ŷ, y) |> x -> x.^2 |> median |> sqrt
maxae(ŷ, y) = err(ŷ, y) |> x -> abs.(x) |> maximum

mape(ŷ, y)    = perr(ŷ, y) |> x -> abs.(x) |> mean
mdape(ŷ, y)   = perr(ŷ, y) |> x -> abs.(x) |> median
rmspe(ŷ, y)   = perr(ŷ, y) |> x -> x.^2 |> mean |> sqrt
rmdspe(ŷ, y)  = perr(ŷ, y) |> x -> x.^2 |> median |> sqrt

"https://en.wikipedia.org/wiki/Symmetric_mean_absolute_percentage_error"
smape(ŷ, y)   = 200. * abs.(err(ŷ, y)) ./ (ŷ + y) |> mean
smdape(ŷ, y)   = 200. * abs.(err(ŷ, y)) ./ (ŷ + y) |> median

# TEST
y    = [-1, 0, 1, 2, 3, 4]
ŷ    = [-3, 1, 2, 2, 4, 5]

e = err(ŷ, y)
p = perr(ŷ, y)

mse(ŷ, y)
rmse(ŷ, y)
mae(ŷ, y) 
mdae(ŷ, y) 
rmdse(ŷ, y)
maxae(ŷ, y)

mape(ŷ, y)
mdape(ŷ, y)
rmspe(ŷ, y)
rmdspe(ŷ, y)
smape(ŷ, y)
smdape(ŷ, y) 
```

**Comment by ablaom** [comment]
Another to add to the list (looking ahead to image classfication) : Tversky Loss
https://github.com/alan-turing-institute/MLJ.jl/issues/554

**Comment by azev77** [comment]
Many are also here: https://github.com/LAMPSPUC/ForecastAccuracy.jl

**Comment by azev77** [comment]
Hey, did anything ever come of this? 
Was this implemented in a different repo?

**Comment by ablaom** [comment]
No, I think the most recent addition of metrics were the multiclass version of FScore and cousins. You can do `measures()` to list what is available.

**Comment by azev77** [comment]
Came across these as well: https://github.com/beacon-biosignals/Lighthouse.jl

**Comment by ablaom** [comment]
Yep. We got this somewhere.

@azev77 I'm currently performing a JuliaAI-wide issue review and collecting all issues in a GH project. All measure-related issues are being given a "measure" label (as this one has) and one can see all these at one glance. I'll make let you know when it's done. Most of these are marked "straightforward" but "low priority" -  but there are a lot of them. I'm trying to see if we can get a student to help out - some of the work is not that hard but it's tedious and not very popular. If you know of someone who we could trust with this work and have a way to pay them...
